### ISSUE TITLE
"cannot open tty-output" on 2nd run of tdm in same tty session

### ISSUE THREAD

**Issue by amakovec** [dopsi/console-tdm#15]
Hello!  First of all, thanks everyone for their work in maintaining console-tdm.  I've read through much of the code, and I believe this project to be a truly robust display manager.

I've been running into a problem with what I think is **dialog**.

**Steps to reproduce:**
1. Log into /dev/tty1.
2. Run tdm.
3. Choose an X session and select "OK".  The X session then initiates successfully.
4. Exit and kill the X session, returning back to /dev/tty1.
5. Run tdm.

This is where the error occurs.  At [console-tdm/tdm:138](https://github.com/dopsi/console-tdm/blob/469a8118225c20f478886db2026e564d2f32cf66/tdm#L138), dialog returns the error **"cannot open tty-output"** and tdm then returns "Falling back to shell".

The odd thing is that I can still run my own dialog commands to stdout (for example, 'dialog --stdout --menu "Test Menu" 0 0 0 0 "Test Item 0" 1 "Test Item 1"').  I've tried debugging [console-tdm/tdm](https://github.com/dopsi/console-tdm/blob/469a8118225c20f478886db2026e564d2f32cf66/tdm) but to no avail.  Maybe some of the contributors have some insight to this problem, as I'm sure you all have a better grasp on the script than I.  Any help is appreciated.

Thanks!

**Comment by dopsi** [comment]
Thanks for the issue report. It should be fixed in [v1.2.3](https://github.com/dopsi/console-tdm/releases/tag/v1.2.3) (2f11771)

**Comment by amakovec** [comment]
Hey @dopsi !

Unfortunately, it doesn't look like this resolved the issue.  Actually, now when I launch `tdm` in `/dev/tty1`, I don't get any output.  I assume this is because of the removal of the `--stdout` flag in [v1.2.3](https://github.com/dopsi/console-tdm/commit/2f117712995647a02b55a5e4317407d476eb73cd), as `dialog` now just writes to the subshell of the command substitution.

I actually figured out a fix to the issue I documented above.  If you could, take a look at my forked repo [amakovec/console-tdm](https://github.com/amakovec/console-tdm/commit/e896090d16053117411a8dc997399b15f8544bdb), I removed the command substitution and replaced it with `sid=$?` after the dialog command.  I actually am not particularly fond of my fix, since it doesn't guarantee that the output of dialog is what will show up in `$?`, but it's a "working prototype", so to speak.

**Comment by dopsi** [comment]
Could you please create a pull request ?

**Comment by amakovec** [comment]
Hey @dopsi !

I've just submitted a pull request with the bugfix [("Fix tdm relaunch after killing X.")](https://github.com/dopsi/console-tdm/pull/16).

My apologies for the delay.  It turns out that my original bugfix that I outlined above ([amakovec/console-tdm/e896090](https://github.com/amakovec/console-tdm/commit/e896090d16053117411a8dc997399b15f8544bdb)) did not fix the bug.  I was assuming that `dialog` was writing the user's choice to the exit status, whereas it actually writes it to `stderr` by default.  I've made a bugfix that redirects `stderr` to a temporary file, where `$sid` then reads from the temporary file to obtain the user's choice.  This keeps `dialog` from locking up `stdout` and allows tdm to be relaunched in the same `tty`.  Just to note, this temporary file method is the same method used in the examples of the [`dialog` source](http://invisible-island.net/dialog/#download).
